### PR TITLE
feat: Add card paring check in onboarding

### PR DESCRIPTION
**Pull request by ujjwal-cyph**
Fixes https://app.clickup.com/t/9002019994/PRF-4760

### CODE REVIEW

File: src/card_operations/check_pairing.c [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+  LOG_SWV("%s (%d): %p\n", __func__, __LINE__, result);
+  card_error_type_e status = CARD_OPERATION_DEFAULT_INVALID;
+  if (NULL == result) {
+    return status;
+  }
+
+  memset(result, 0, sizeof(check_pairing_result_t));
+  result->card_number = 0;
+  card_operation_data_t operation_data = {tap_card_data, NULL, 0};
+  init_tap_card_data(&operation_data.nfc_data);
```
Comment (amanCypherock) [inline src/card_operations/check_pairing.c:126]: Here we are first initializing with tap_card_data then resetting it. Is tap_card_data left by mistake? As the nfc_data member is not a pointer

File: src/card_operations/check_pairing.c [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    }
+  } else {
+    if (NULL != operation_data.error_message) {
+      delay_scr_init(operation_data.error_message, DELAY_TIME);
+    }
+    LOG_SWV("%s (%d):", __func__, __LINE__);
+    LOG_CRITICAL("COCP %04X\n", operation_data.nfc_data.status);
+  }
+  buzzer_start(100);
+  return status;
```
Comment (amanCypherock) [inline src/card_operations/check_pairing.c:146]: On return, please call nfc_deselect_card to turn off the NFC session.

File: src/card_operations/check_pairing.c [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+/*****************************************************************************
+ * PRIVATE TYPEDEFS
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTION PROTOTYPES
+ *****************************************************************************/
+
+/**
+ * TODO: add description when this api gets concrete definition
```
Comment (amanCypherock) [inline src/card_operations/check_pairing.c:85]: Use GPT :)

File: apps/manager_app/card_training.c [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+void manager_card_training(manager_query_t *query) {
+  LOG_SWV("%s (%d)\n", __func__, __LINE__);
+  if (NULL == query) {
+    // TODO: verify card training query tag for completeness
+    return;
+  }
+  // TODO: verify on-boarding state for safety
+  core_status_set_device_waiting_on(CORE_DEVICE_WAITING_ON_BUSY_IP_CARD);
+
+  instruction_scr_init("Place any card below the device to test card tapping",
```
Comment (amanCypherock) [inline apps/manager_app/card_training.c:145]: move to constant text

File: apps/manager_app/card_training.c [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+  // always pair the card
+  // TODO: Update the pairing flow to support flexible storage of shared key
+
+  // TODO: Fetch wallet list; send dummy data
+  result.result.card_paired = false;
+  result.result.wallet_list_count = 0;
+  send_message_to_host(&result);
+
+  char msg[64] = "";
+  snprintf(msg, sizeof(msg), "Card #%d Tapped", pair_result.card_number);
```
Comment (amanCypherock) [inline apps/manager_app/card_training.c:168]: Use strings from constant texts

File: apps/manager_app/card_training.c [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+  delay_scr_init(msg, DELAY_TIME);
```
Comment (amanCypherock) [inline apps/manager_app/card_training.c:169]: How will we synchronise this delay with Cysync?

Comment (ujjwal-cyph) [reply apps/manager_app/card_training.c:169]: mutual understanding between them

File: src/card_operations/check_pairing.c [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        decode_card_number(operation_data.nfc_data.tapped_card);
+    if (-1 == is_paired(operation_data.nfc_data.card_key_id)) {
+      // keystore index is -1; the keystore does not have key-id
+      result->is_paired = false;
+    } else {
+      result->is_paired = true;
+    }
+  } else {
+    if (NULL != operation_data.error_message) {
+      delay_scr_init(operation_data.error_message, DELAY_TIME);
```
[the remark this answers is not in the document]
Comment (ujjwal-cyph) [reply src/card_operations/check_pairing.c:140]: did not have to handle error cases. optimizing for only the happy flow. this is a temporary implementation

File: apps/manager_app/card_training.c (hunk 6 above)
Comment (ujjwal-cyph) [reply apps/manager_app/card_training.c:169]: mutual understanding between them

File: src/card_operations/check_pairing.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (ujjwal-cyph) [reply src/card_operations/check_pairing.c]: changed

File: src/card_operations/check_pairing.c (hunk 2 above)
Comment (ujjwal-cyph) [reply src/card_operations/check_pairing.c:149]: added

File: apps/manager_app/card_training.c (hunk 4 above)
Comment (ujjwal-cyph) [reply apps/manager_app/card_training.c]: changed

File: apps/manager_app/card_training.c (hunk 5 above)
Comment (ujjwal-cyph) [reply apps/manager_app/card_training.c]: changed

File: src/card_operations/check_pairing.c (hunk 1 above)
Comment (ujjwal-cyph) [reply src/card_operations/check_pairing.c]: removed

File: src/card_operations/check_pairing.c [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+  memset(result, 0, sizeof(check_pairing_result_t));
+  result->card_number = 0;
+  card_operation_data_t operation_data = {{0}, NULL, 0};
+  init_tap_card_data(&operation_data.nfc_data);
+  status = card_initialize_applet(&operation_data);
+  buzzer_start(100);
+  nfc_deselect_card();
+
+  if (CARD_OPERATION_SUCCESS == status) {
```
Comment (kartikCypherock) [inline src/card_operations/check_pairing.c:132]: Please unify error handling as done in card authentication by Aman, once that is merged

Comment (ujjwal-cyph) [reply src/card_operations/check_pairing.c:132]: there are no cases that need retrying here. card authentication has some error scenarios where retrying is feasible. such as wrong card sequence, wrong family-id, etch. such cases will not happen in this particular flow.

**Review by kartikCypherock — APPROVED** [review]

File: apps/manager_app/card_training.c [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+  manager_train_card_response_t result = MANAGER_TRAIN_CARD_RESPONSE_INIT_ZERO;
+  result.which_response = MANAGER_TRAIN_CARD_RESPONSE_RESULT_TAG;
+  card_error_type_e status = card_check_pairing(&pair_result);
+  if (CARD_OPERATION_SUCCESS != status) {
+    LOG_SWV("%s (%d)\n", __func__, __LINE__);
+    send_training_error(1);
+    return;
+  }
+  result.result.card_paired = pair_result.is_paired;
+
```
Comment (amanCypherock) [inline apps/manager_app/card_training.c:157]: Here, we should also set the family id to flash, as we expect to follow the next step of onboarding, i.e. Card Auth with the same family id.

Comment (ujjwal-cyph) [reply apps/manager_app/card_training.c:157]: added

**Review by amanCypherock — APPROVED** [review]

**Pull request merged**